metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,569 @@
+      indentation = getIndentationForPositionInDocument(applyTree.pos, document)
+    } yield {
+      applyTree match {
+        case termApply: Term.Apply =>
+          maybeBuildToForYieldChainedCodeActionWithApply(
+            path,
+            termApply,
+            document,
+            indentation
+          )
+        case termSelect: Term.Select
+            if termSelect.name.value == "map" || termSelect.name.value == "flatMap" ||
+              termSelect.name.value == "filter" || termSelect.name.value == "filterNot" =>
+          maybeBuildToForYieldCodeActionWithSelect(
+            path,
+            termSelect,
+            document,
+            indentation
+          )
+
+        case termName: Term.Name
+            if termName.value == "flatMap" || termName.value == "map" ||
+              termName.value == "filter" || termName.value == "filterNot" =>
+          maybeBuildToForYieldCodeActionWithName(
+            path,
+            termName,
+            document,
+            indentation
+          )
+
+        case _ => None
+      }
+    }
+
+    maybeChainedCodeAction.flatten.toSeq
+
+  }
+
+  private def buildCodeActionFromForYieldParts(
+      nameQualsList: List[ForYieldEnumeration],
+      forYieldConditionsList: List[ForYieldCondition],
+      yieldExpression: YieldExpression,
+      indentation: String,
+      path: AbsolutePath,
+      startPos: l.Position,
+      endPos: l.Position
+  ): l.CodeAction = {
+
+    val enumerations = nameQualsList.flatMap { nameQual =>
+      for {
+        valName <- nameQual.perhapsVariableName
+        assignOrMap <- nameQual.perhapsAssignOrMap
+        qual <- nameQual.qual
+      } yield (valName, assignOrMap, qual)
+    }
+
+    val conditions = forYieldConditionsList.flatMap { condition =>
+      for {
+        conditionString <- condition.condition
+        filterOrNot <- condition.maybeFilterOrNot
+      } yield (filterOrNot, conditionString)
+
+    }
+
+    val forYieldString =
+      s"""|for {
+          |${enumerations.map(nameQual => s"${indentation}  ${nameQual._1} ${nameQual._2} ${nameQual._3}").mkString("\n")}
+          |${conditions.map(forYieldCondition => s"${indentation}  if ${forYieldCondition._1}(${forYieldCondition._2})").mkString("\n")}
+          |${indentation}} yield {
+          |${indentation}  ${yieldExpression.expression}
+          |${indentation}}""".stripMargin
+
+    val codeAction = new l.CodeAction()
+    val range =
+      new l.Range(startPos, endPos)
+    codeAction.setTitle(
+      FlatMapToForComprehensionCodeAction.flatMapToForComprehension
+    )
+    codeAction.setKind(this.kind)
+    val forComprehensionTextEdit = new l.TextEdit(range, forYieldString)
+    codeAction.setEdit(
+      new l.WorkspaceEdit(
+        Map(
+          path.toURI.toString -> List(forComprehensionTextEdit).asJava
+        ).asJava
+      )
+    )
+    codeAction
+  }
+
+  private def maybeBuildToForYieldChainedCodeActionWithApply(
```

Comment:
```suggestion
  private def forComprehensionAction(
```
it's more difficult to see differences between names if they are long, so try to keep the names as different as possible to be able to distinguish different functions.